src/json.hpp.re2c

```diff
@@ -8209,177 +8209,338 @@ class basic_json
+                while (p < m_end and *p >= '0' 
+                                 and *p <= '9') 
+                {
+                    ++p;
+                }
 
-            const lexer::lexer_char_t* curptr = m_start;
+                return p == m_end;
+            }
 
-            // accumulate the integer conversion result (unsigned for now)
-            number_unsigned_t value = 0;
+        private:
+            const char* const m_start = nullptr;
+            const char* const m_end  = nullptr;
 
-            // maximum absolute value of the relevant integer type
-            number_unsigned_t max;
+            // overloaded wrappers for strtod/strtof/strtold
+            // that will be called from parse<floating_point_t>
 
-            // temporarily store the type to avoid unecessary bitfield access
-            value_t type;
+            static void strtof(float& f,
+                               const char* str,
+                               char** endptr)
+            {
+                f = std::strtof(str, endptr);
+            }
 
-            // look for sign
-            if (*curptr == '-')
+            static void strtof(double& f,
+                               const char* str,
+                               char** endptr)
             {
-                type = value_t::number_integer;
-                max = static_cast<uint64_t>((std::numeric_limits<number_integer_t>::max)()) + 1;
-                curptr++;
+                f = std::strtod(str, endptr);
             }
-            else
+
+            static void strtof(long double& f,
+                               const char* str,
+                               char** endptr)
             {
-                type = value_t::number_unsigned;
-                max = static_cast<uint64_t>((std::numeric_limits<number_unsigned_t>::max)());
+                f = std::strtold(str, endptr);
             }
 
-            // count the significant figures
-            for (; curptr < m_cursor; curptr++)
+            template<typename T>
+            bool parse(T& value, /*is_integral=*/std::false_type) const
             {
-                // quickly skip tests if a digit
-                if (*curptr < '0' || *curptr > '9')
-                {
-                    if (*curptr == '.')
+                const char* data = m_start;
+                const size_t len = static_cast<size_t>(m_end - m_start);
+
+                const char decimal_point_char =
+                    std::use_facet< std::numpunct<char> >(
+                        std::locale()).decimal_point();
+
+                // replace decimal separator with locale-specific
+                // version, when necessary; data will be repointed
+                // to either buf or tempstr containing the fixed
+                // string.
+                std::string tempstr;
+                std::array<char, 64> buf;
+                do {
+                    if (decimal_point_char == '.') 
```

Comment:
`setlocale` with C locales is the equivalent of `std::locale::global` with C++ locales, which is not thread-specific or thread-safe either. 

We're not calling it it from the library code, however, but from the unit test to simulate the current state of a locale-specific application. All locale-dependent functions, including `std::to_string()` or `printf()` or `cout <<` are not thread-safe in that sense, as a rogue thread may change C or C++ locale from under your feet by changing the locale in-flight because "they" needed it.